ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/Pred.hs

```diff
@@ -0,0 +1,246 @@
+  Bool :: !Bool -> Pred tag
+  -- | Boolean negation
+  Not :: !(Pred tag) -> Pred tag
+  -- | Greater-than comparison
+  (:>:) :: (Typeable a, Ord a, Show a) => !a -> !a -> Pred tag
+  -- | Less-than-or-equal comparison
+  (:<=:) :: (Typeable a, Ord a, Show a) => !a -> !a -> Pred tag
+  -- | Equality comparison
+  (:==:) :: (Typeable a, Eq a, Show a) => !a -> !a -> Pred tag
+  -- | Conjunction
+  (:/\:) :: !(Pred tag) -> !(Pred tag) -> Pred tag
+  -- | Disjunction
+  (:\/:) :: !(Pred tag) -> !(Pred tag) -> Pred tag
+
+deriving instance Show tag => Show (Pred tag)
+
+instance Eq tag => Eq (Pred tag) where
+  (t1 := p1) == (t2 := p2) =
+    t1 == t2 && p1 == p2
+  Bool b1 == Bool b2 =
+    b1 == b2
+  Not p1 == Not p2 =
+    p1 == p2
+  (a1 :>: b1) == (a2 :>: b2)
+    | Just (a2', b2') <- cast (a2, b2) =
+        a1 == a2' && b1 == b2'
+  (a1 :<=: b1) == (a2 :<=: b2)
+    | Just (a2', b2') <- cast (a2, b2) =
+        a1 == a2' && b1 == b2'
+  (a1 :==: b1) == (a2 :==: b2)
+    | Just (a2', b2') <- cast (a2, b2) =
+        a1 == a2' && b1 == b2'
+  (a1 :/\: b1) == (a2 :/\: b2) =
+    a1 == a2 && b1 == b2
+  (a1 :\/: b1) == (a2 :\/: b2) =
+    a1 == a2 && b1 == b2
+  _ == _ =
+    False
+
+infixr 2 :=
+
+infixr 3 :\/:
+
+infixr 4 :/\:
+
+infixr 5 `Not`
+
+infix 5 :>:
+
+infix 5 :<=:
+
+infix 5 :==:
+
+-- | Sufficient evidence to show that a predicate is either true or false
+data Evidence (res :: Bool) a where
+  ETrue :: Pred a -> Evidence True a
+  EFalse :: Pred a -> Evidence False a
+
+deriving instance Show a => Show (Evidence res a)
+deriving instance Eq a => Eq (Evidence res a)
+
+-- | Evaluate a predicate, yielding either a counterexample or a witness.
+--
+-- The returned value contains the minimum (modulo conjunction/disjunction
+-- short circuiting) evidence needed to explain the outcome.
+--
+-- Some examples:
+--
+-- >>> data P = A | B | C deriving Show
+-- >>> a = A := Bool True   -- a ~ True
+-- >>> b = B := 2+2 :==: 5  -- b ~ False
+-- >>> c = C := 10 :>: 5    -- c ~ True
+--
+-- >>> evalPred (a :/\: c) show  -- success because both a~True and c~True
+-- ETrue "((A := Bool True) :/\: (C := 10 :>: 5))"
+--
+-- >>> evalPred (a :\/: b) show  -- success because a~True, short-circuits
+-- ETrue "(A := Bool True)"
+--
+-- >>> evalPred (a :/\: b :/\: c) show  -- failure because b~False, short-circuits
+-- EFalse "(B := 4 :==: 5)"
+--
+-- >>> evalPred ((b :\/: a) :/\: (b :\/: c)) show  -- success because of a~True and c~True
+-- ETrue "((A := Bool True) :/\: (C := 10 :>: 5))"
+--
+-- >>> evalPred (b :\/: (Not c)) show  -- failure because both b~False and c~True
+-- EFalse "((B := 4 :==: 5) :\/: Not (C := 10 :>: 5))"
+--
+-- >>> evalPred (Not (a :/\: b)) show  -- success because b~False
+-- ETrue "(Not (B := 4 :==: 5))"
+--
+-- >>> evalPred (Not (a :/\: c)) show  -- failure because both a~True and c~True
+-- EFalse "(Not ((A := Bool True) :/\: (C := 10 :>: 5)))"
+evalPred :: Pred tag -> (forall res. Evidence res tag -> r) -> r
```

Comment:
I have no strong opinion on this, so feel free to ignore, but we could have kept the `Either`, right?

Is the CPS-style adding any benefit?

Reply:
Using CPS is the (IMO) standard way to introduce existential types. 

FTR: using 

```haskell
evalPred :: Pred tag -> Either (Evidence True tag) (Evidence False tag)
```
also works, but breaks in a more annoying way if we want to add a new constructor to `Evidence` in the future (because we either need to introduce a three-way either, or to bite the bullet an use CPS). Adding a new constructor when you have everything already written using CPS is cleaner, as we just get warnings about non-exhaustive case expressions.